examples/vision/deit.py

```diff
@@ -0,0 +1,575 @@
+
+        with tf.GradientTape() as tape:
+            # Forward pass of student.
+            cls_predictions, dist_predictions = self.student(x / 255.0, training=True)
+
+            # Compute losses.
+            student_loss = self.student_loss_fn(y, cls_predictions)
+            distillation_loss = self.distillation_loss_fn(
+                teacher_predictions, dist_predictions
+            )
+            loss = (student_loss + distillation_loss) / 2
+
+        # Compute gradients.
+        trainable_vars = self.student.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+
+        # Update weights.
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Update the metrics configured in `compile()`.
+        student_predictions = (cls_predictions + dist_predictions) / 2
+        self.compiled_metrics.update_state(y, student_predictions)
+
+        # Return a dict of performance.
+        results = {m.name: m.result() for m in self.metrics}
+        results.update(
+            {"student_loss": student_loss, "distillation_loss": distillation_loss}
+        )
+        return results
+
+    def test_step(self, data):
+        # Unpack the data.
+        x, y = data
+
+        # Compute predictions.
+        y_prediction = self.student(x / 255.0, training=False)
+
+        # Calculate the loss.
+        student_loss = self.student_loss_fn(y, y_prediction)
+
+        # Update the metrics.
+        self.compiled_metrics.update_state(y, y_prediction)
+
+        # Return a dict of performance.
+        results = {m.name: m.result() for m in self.metrics}
+        results.update({"student_loss": student_loss})
+        return results
+
+    def call(self, inputs):
+        return self.student(inputs / 255.0, training=False)
+
+
+"""
+## Load the teacher model
+
+This model is based on the BiT family of ResNets
+([Kolesnikov et al.](https://arxiv.org/abs/1912.11370))
+fine-tuned on the `tf_flowers` dataset. You can refer to
+[this notebook](https://github.com/sayakpaul/deit-tf/blob/main/notebooks/bit-teacher.ipynb)
+to know how the training was performed. The teacher model has about 212 Million parameters
+which is about **40x more** than the student.
+"""
+
+"""shell
+wget -q https://github.com/sayakpaul/deit-tf/releases/download/v0.1.0/bit_teacher_flowers.zip
+unzip -q bit_teacher_flowers.zip
+"""
+
+# teacher_model_file = keras.utils.get_file(
+#     origin="https://github.com/sayakpaul/deit-tf/releases/download/v0.1.0/bit_teacher_flowers.zip",
+#     extract=True,
+#     cache_subdir="models"
+# )
+bit_teacher_flowers = keras.models.load_model("bit_teacher_flowers")
+
+"""
+## Training through distillation
+"""
+
+deit_tiny = ViTDistilled()
+deit_distiller = DeiT(student=deit_tiny, teacher=bit_teacher_flowers)
+
+lr_scaled = (BASE_LR / 512) * BATCH_SIZE
+deit_distiller.compile(
+    optimizer=tfa.optimizers.AdamW(weight_decay=WEIGHT_DECAY, learning_rate=lr_scaled),
```

Comment:
Fun fact: AdamW has recently been included in the API (in Keras 2.9 it's at `keras.optimizers.experimental.AdamW` and it will be at `keras.optimizers.AdamW` likely in 2.10). But let's keep TFA for now.